src/functional/iterator/embedded.py

```diff
@@ -845,29 +845,26 @@ def get_ordered_indices(
     return tuple(res)
 
 
-def _tupsum(a, b):
-    def combine_slice(s, t):
-        is_slice = False
-        if isinstance(s, slice):
-            is_slice = True
-            first = 0 if s.start is None else s.start
-            assert s.step is None
-            assert s.stop is None
-        else:
-            assert isinstance(s, numbers.Integral)
-            first = s
-        if isinstance(t, slice):
-            is_slice = True
-            second = 0 if t.start is None else t.start
-            assert t.step is None
-            assert t.stop is None
-        else:
-            assert isinstance(t, numbers.Integral)
-            second = t
-        start = first + second
-        return slice(start, None) if is_slice else start
+def _shift_slice(
+    slice_or_index: slice | numbers.Integral, offset: numbers.Integral, size: numbers.Integral
+) -> slice | numbers.Integral:
+    if isinstance(slice_or_index, slice):
+        assert slice_or_index.step is None
+        return slice(
+            0 if slice_or_index.start is None else slice_or_index.start + offset,
+            None if slice_or_index.stop is None else min(slice_or_index.stop + offset, size),
+        )
```

Comment:
```suggestion
        start = slice_or_index.start or 0
        stop = slice_or_index.stop or 0
        # avoid unintentional wrap around
        assert sign(start+offset) == sign(start) and sign(stop+offset) == sign(stop)
        return slice(start + offset, stop+offset)
```
Why is the offset not respected if the start or stop are None? Additionally the `min` appears to be intended for protecting against taking a too large slice. Shouldn't this actually throw an error?

Reply:
Thanks for this comment. I realized how messed-up the current implementation was: the slice that is passed here has the semantics of a range relative to the origin of a field, i.e. negative indices don't mean relative to the end.
I changed the behavior of  `__getitem__`/ `__setitem__` to operate on the full buffer, i.e. not relative to the origin and introduced 2 new functions that are used internally with the semantics relative to the origin. I'll test this now in icon4py.